### DISCUSSION TITLE
Save and restore MainWindow state, size and position

### DISCUSSION THREAD

**Discussion by Sputnik24** [AvaloniaUI/Avalonia#7836] [Q&A]
How can I save the Window's state (normal or maximized), size and position and restore them on startup?

**Comment by robloo** [comment]
For size: Have you tried using the [Window.Height](http://reference.avaloniaui.net/api/Avalonia.Layout/Layoutable/C5E69863) and [Window.Width](http://reference.avaloniaui.net/api/Avalonia.Layout/Layoutable/6501B150) properties? There is also [Window.Bounds](http://reference.avaloniaui.net/api/Avalonia/Visual/8E0A265B).

You would need to save these before shutdown so will need to monitor the event. https://stackoverflow.com/questions/66920381/how-to-handle-system-shutdown-event-using-avalonia-framework-on-both-windows-and

Related:
 - https://github.com/AvaloniaUI/Avalonia/issues/5285
 - https://github.com/AvaloniaUI/Avalonia/issues/5005

**Reply by LazaroOnline** [reply]
If only avalonia had a property like "persistWindowState" that just made it work out of the box, or id there was a nuget package to do this in one line...

**Reply by thevortexcloud** [reply]
There is nothing stopping you from writing an extension method to save and restore the state. However it's far more complicated than it sounds. EG on desktops there might be several monitors, if the window is placed on a monitor and then the monitor is disconnected, when you go to restore the window it will probably be off screen on some platforms. Similarly the resolution of a monitor can also change, as such the position it was in may no longer be possible.

**Reply by LazaroOnline** [reply]
Turns out there is an awesome generic NuGet Package to do this very easily here:
https://github.com/anakic/Jot
It has examples for WFP and WinForms, and I just added an example for AvaloniaUI in the Readme:
https://github.com/anakic/Jot?tab=readme-ov-file#avalonia-ui

However JOT only takes care of persisting values, it doesn't have logic to "determine if a window is within the screen boundaries", which should have into account multiple monitors, windows that are half in / half out, windows 11 docking system using position "-1", but specially to check if at least some part of the "draggable" window title bar is visible within any of the screens.
It would be nice if Avalonia had a function to get this information, something like "`isWindowDraggableVisible()`", or if that is too much then at least "`isWindowVisible()`".

**Reply by SimonORorke** [reply]
> There is nothing stopping you from writing an extension method to save and restore the state. However it's far more complicated than it sounds. EG on desktops there might be several monitors, if the window is placed on a monitor and then the monitor is disconnected, when you go to restore the window it will probably be off screen on some platforms. Similarly the resolution of a monitor can also change, as such the position it was in may no longer be possible.

That's an excellent specification of the requirements. In the past, I've implemented this for Windows Forms and WPF. I'm contemplating doing the same for Avalonia, but I don't know what cross platform issues there might be.  I'm developing in Windows but also targeting macOS.

IMHO, the complexity is all the more reason why it would be great if Avalonia were to provide this facility out of the box. It _is_ provided by JUCE, a C++ cross platform framework. JUCE's architecture is very different from Avalonia; and JUCE's facility assumes that the variables are to be persisted in a settings file of a standard type they provide.  In Avalonia, I envisage a facility that does all that logic to do with screen variability and presents results that the developer can persist in whatever settings storage type is required.

**Reply by thevortexcloud** [reply]
Avalonia does already have a way to query screen information. However I doubt it will ever implement this functionality itself, simply because it's not possible on every platform. Wayland (what almost every desktop on Linux is moving to) for example does not allow for windows to be absolute positioned (or even position themselves). There is a lot of discussion about Wayland allowing some sort of self window positioning, but they are trying very hard to avoid the whole window coordinate system. EG 

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/264

The primary reason is most window coordinate systems assume a square display. While Wayland is trying to support every possible display shape, and even non physical displays (think a VR desktop, which would have an X, Y and Z coordinate system compared to the X, Y of a classic desktop). Which means Wayland will likely never allow this beyond providing some basic hint to the window manager about what to do. 

This is possible however on X11 on Linux as X11 is fairly old and handles Window placement a lot more like Windows. However you may still run into problems on Linux as people like to use tiling window manager, which will try to place windows in specific locations with specific sizes, regardless of what the window wants to do.

> but also targeting macOS.

Unfortunately I don't know enough about MacOS to comment on it. On windows however this should be possible. You are very likely going to be on your own if you need/want this functionality. It's a very hard problem to solve in a reliable way cross platform. Not even Windows itself can get it right when you disconnect an external display on a laptop in most cases.

**Reply by SimonORorke** [reply]
Thanks, @thevortexcloud.

> Avalonia does already have a way to query screen information.

Excellent, that's what I was hoping.

> It's a very hard problem to solve in a reliable way cross platform. Not even Windows itself can get it right when you disconnect an external display on a laptop in most cases.

Yes, my experience with implementing this for Windows Forms and WPF was that such hazards make it impossible to 100% reliably save sensible window size/position/state settings.  The key, I found, is to allow for persisted settings that don't make sense in the new screen environment when the application is next loaded.

Fortunately, the application I'm developing is very unlikely to be wanted on Linux.  I only require Windows and macOS.

I shall have a go at this. But, if anyone can tell me about any macOS-specific considerations I should know about, I would appreciate it.

**Reply by thevortexcloud** [reply]
> Excellent, that's what I was hoping.

I should note the information it does give is fairly minimal. The `Screen` class in Avalonia is directly based on the Winforms one. Although I think it should give you everything you need?

https://reference.avaloniaui.net/api/Avalonia.Controls/Screens/

The biggest limitation I can think of is it returns a list of screens, with no real indication what each screen is beyond display bounds/position. So if you have multiple identical screens and one is disconnected you have no real way to tell which was which. In which case you may need to resort to a third party library like SDL. Which lets you query some extra EDID information as well as significantly more display information. EG you can get the actual display name https://wiki.libsdl.org/SDL2/SDL_GetDisplayName

There is a pretty good C# binding for SDL2 as well.

https://github.com/flibitijibibo/SDL2-CS

Although you probably want the actual display serial number, which EDID does provide. But SDL2 does not from what I can tell. Querying EDID directly on Windows is not straightforward from what I know. On Linux it's just a file you can read.

**Reply by SimonORorke** [reply]
Thanks again, @thevortexcloud. Yes, it does look like the Avalonia `Screens` and `Screen` classes should tell me what I need.  Judging by my Windows Forms implementation (I don't have what I wrote for WPF), I should be able to get away without having to identify specific monitors by name or serial.  I shall investigate the SLD2 links you provided if need be. If I get something to work, I shall post some code here.

**Comment by SimonORorke** [comment]
Here's my solution. I've only tested it on Windows. Judging from the preceding discussion, it will probably also work on macOS and Linux with X11.
```csharp
/// <summary>
///   A service that facilitates persistence of the main window's position, size and
///   state.
/// </summary>
/// <remarks>
///   This service does not actually read/save settings to/from storage. That is taken
///   care of in the view model. 
/// </remarks>
public class WindowLocationService : IWindowLocationService {
  private Window? _mainWindow;
  private Window MainWindow => _mainWindow ??= ((App)Application.Current!).MainWindow;
  public int? Left { get; set; }
  public int? Top { get; set; }
  public int? Width { get; set; }
  public int? Height { get; set; }
  public int? WindowState { get; set; }

  private Screen? FindScreenContainingPositionInWorkingArea(PixelPoint position) {
    return (
      // All active screens, not just any screens overlapping the window! 
      from screen in MainWindow.Screens.All 
      where screen.WorkingArea.Contains(position)
      select screen).FirstOrDefault();
  }

  /// <summary>
  ///  Uses previously saved data, if available, to size the window and position it on a
  ///  screen. 
  /// </summary>
  public void Restore() {
    // If the settings have not previously been saved, they should all be null;
    // once saved, none should.
    if (this is not {
          Left: not null, Top: not null, Width: not null, Height: not null,
          WindowState: not null
        }) {
      // There are no settings to restore.
      // So leave the windows size and position at their defaults.
      return;
    }
    var savedPosition = new PixelPoint(Left.Value, Top.Value);
    var screen = FindScreenContainingPositionInWorkingArea(savedPosition);
    if (screen == null) {
      // The saved window position (its top left corner) is not in the working area of
      // an active screen.
      // So leave the windows size and position at their defaults.
      return;
    }
    const int min = 50;
    if (Left.Value > screen.WorkingArea.X + screen.WorkingArea.Width - min
        || Top.Value > screen.WorkingArea.Y + screen.WorkingArea.Height - min) {
      // The saved top left corner (position) is so close to the right or bottom edge
      // of the screen's working area as to make the make the window difficult to access.
      // So leave the windows size and position at their defaults.
      return;
    }
    MainWindow.Position = savedPosition;
    MainWindow.Width = Width.Value;
    MainWindow.Height = Height.Value;
    MainWindow.WindowState = (WindowState)WindowState.Value;
  }

  /// <summary>
  ///   Updates the data to be saved to settings from the window's current
  ///   position/size/state.
  /// </summary>
  public void Update() {
    switch (MainWindow.WindowState) {
      case Avalonia.Controls.WindowState.Minimized:
        // We don't want to reopen the window minimized. And the window's position
        // properties (X and Y) are junk when minimised. So we must avoid saving them.
        return;
      case Avalonia.Controls.WindowState.Normal:
        Left = MainWindow.Position.X;
        Top = MainWindow.Position.Y;
        Width = (int)MainWindow.Width;
        Height = (int)MainWindow.Height;
        break;
    }
    // When the window is maximised, its position properties (X and Y) are for the top
    // left corner of the screen. We have avoided saving them so that, if the user
    // subsequently restores the window to its normal position, it will revert to the
    // previously saved normal position rather than been being put in the top left
    // corner.
    WindowState = (int)MainWindow.WindowState;
  }
}
```
Usage in the view would be something like this:
```csharp
public class MainWindow : Window {
  public MainWindow() {
    InitializeComponent();
  }

  protected override void OnClosing(WindowClosingEventArgs e) {
      ViewModel.WindowLocationService.Update();
      ViewModel.Close(); // View model will save the settings to storage.
  }

  protected override void OnLoaded(RoutedEventArgs e) {
    ViewModel = (MainWindowViewModel)DataContext!;
    // View model has read the settings from storage.
    ViewModel.WindowLocationService.Restore();
  }
}
```
For the interface, view model, service registration etc., see https://github.com/SimonORorke/FalconProgrammer.
